snapcraft/internal/lxd.py

```diff
@@ -139,6 +145,77 @@ def _setup_project(self):
         self._push_file(tar_filename, dst)
         self._container_run(['tar', 'xvf', os.path.basename(tar_filename)])
 
+    def _inject_snapcraft(self):
+        if common.is_snap():
+            # Because of https://bugs.launchpad.net/snappy/+bug/1628289
+            self._container_run(['apt-get', 'install', 'squashfuse', '-y'])
+
+            # Push core snap into container
+            self._inject_snap('core')
+            self._inject_snap('snapcraft')
+        else:
+            self._container_run(['apt-get', 'install', 'snapcraft', '-y'])
+
+    def _inject_snap(self, name):
+        session = requests_unixsocket.Session()
+        snapd_socket = '/run/snapd.socket'.replace('/', '%2F')
+        # Cf. https://github.com/snapcore/snapd/wiki/REST-API#get-v2snapsname
+        api = 'http+unix://{}/v2/snaps/{}'.format(snapd_socket, name)
+        try:
+            json = session.request('GET', api).json()
+        except requests.exceptions.ConnectionError as e:
+            raise SnapcraftEnvironmentError(
+                'Error connecting to {}'.format(api)) from e
+        if json['type'] == 'error':
+            raise SnapcraftEnvironmentError(
+                'Error querying {!r} snap: {}'.format(
+                    name, json['result']['message']))
+        id = json['result']['id']
+        # Lookup confinement to know if we need to --classic when installing
+        classic = json['result']['confinement'] == 'classic'
+        # Revisions are unique, so we don't need to know the channel
+        rev = json['result']['revision']
+
+        # Copy files to a path the 'lxd' snap can access
+        rundir = os.path.join(os.path.sep, 'run', 'user', str(os.getuid()),
+                              'snap.lxd')
+        os.makedirs(rundir, exist_ok=True)
+
+        self._inject_assertions(rundir, '{}_{}.assert'.format(name, rev), [
+            ['account-key', 'public-key-sha3-384={}'.format(_STORE_KEY)],
+            ['snap-declaration', 'snap-name={}'.format(name)],
+            ['snap-revision', 'snap-revision={}'.format(rev),
+             'snap-id={}'.format(id)],
+        ])
+
+        # https://github.com/snapcore/snapd/blob/master/snap/info.go
+        # MountFile
+        filename = '{}_{}.snap'.format(name, rev)
+        # https://github.com/snapcore/snapd/blob/master/dirs/dirs.go
+        # CoreLibExecDir
+        installed = os.path.join(os.path.sep, 'var', 'lib', 'snapd', 'snaps',
+                                 filename)
+
+        filepath = os.path.join(rundir, filename)
+        shutil.copyfile(installed, filepath)
+        self._push_file(filepath, os.path.join(self._project_folder, filename))
+        logger.info('Installing {}'.format(filename))
+        cmd = ['snap', 'install', filename]
+        if classic:
```

Comment:
we probably need a `is_dangerous` to cover for the case where we built our own snapcraft snap